packages/main/src/Popover.js

```diff
@@ -197,202 +233,209 @@ class Popover extends Popup {
 		return litRender;
 	}
 
+	static get styles() {
+		return PopoverCss;
+	}
+
 	static get template() {
 		return PopoverTemplate;
 	}
 
-	static get styles() {
-		return [Popup.styles, popoverCss];
+	static async define(...params) {
+		super.define(...params);
 	}
 
-	constructor() {
-		super();
-
-		this._documentMouseDownHandler = this.documentMouseDown.bind(this);
+	forwardToFirst() {
+		const firstFocusable = FocusHelper.findFirstFocusableElement(this.contentDOM);
 
-		const that = this;
-
-		this._focusElementsHandlers = {
-			forwardToFirst: event => {
-				const firstFocusable = FocusHelper.findFirstFocusableElement(that);
+		if (firstFocusable) {
+			firstFocusable.focus();
+		}
+	}
 
-				if (firstFocusable) {
-					firstFocusable.focus();
-				}
-			},
-			forwardToLast: event => {
-				const lastFocusable = FocusHelper.findLastFocusableElement(that);
+	forwardToLast() {
+		const lastFocusable = FocusHelper.findLastFocusableElement(this.contentDOM);
 
-				if (lastFocusable) {
-					lastFocusable.focus();
-				}
-			},
-		};
+		if (lastFocusable) {
+			lastFocusable.focus();
+		}
 	}
 
-	isModal() {
-		return this.modal;
+	isOpenerClicked(event) {
+		return event.target === this._opener;
 	}
 
-	static isInRect(x, y, rect) {
-		return x >= rect.left && x <= rect.right
-			&& y >= rect.top && y <= rect.bottom;
-	}
+	openBy(opener) {
+		if (!opener || this.opened) {
+			return;
+		}
 
-	static getClientRect(domRef) {
-		const rect = domRef.getBoundingClientRect();
-		const computedStyle = window.getComputedStyle(domRef);
+		// close all popovers untill we find that target is contained in some of them
+		while (getRegistry().length && (!getRegistry()[getRegistry().length - 1].contains(opener))) {
+			getRegistry()[getRegistry().length - 1].close();
+		}
 
-		const offsetLeft = parseFloat(computedStyle.paddingLeft);
-		const offsetRight = parseFloat(computedStyle.paddingRight);
-		const offsetTop = parseFloat(computedStyle.paddingTop);
-		const offsetBottom = parseFloat(computedStyle.paddingBottom);
+		this._opener = opener;
+		this._focusedElementBeforeOpen = getFocusedElement();
 
-		return {
-			left: rect.left + offsetLeft,
-			right: rect.right - offsetRight,
-			top: rect.top + offsetTop,
-			bottom: rect.bottom - offsetBottom,
-			width: rect.width - offsetLeft - offsetRight,
-			height: rect.height - offsetTop - offsetBottom,
-		};
-	}
+		this.fireEvent("beforeOpen", {});
+		this.reposition();
+		this.applyInitialFocus();
 
-	hitTest(event) {
-		const domRef = this.getPopupDomRef();
-		const rect = domRef.getBoundingClientRect();
-		let x,
-			y;
+		addOpenedPopover(this);
```

Comment:
Two lines above there is reposition() called, and addOpenedPopover(this) itself will call reposition() (though runUpdateInterval). Is this intended?

Reply:
I wanted to call it synchronously so I can be sure that onbefore / after open will be called in the correct order.